src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs

```diff
@@ -239,15 +246,19 @@ private int StartTestRunPrivate(ITestRunEventsHandler runEventsHandler)
             this.currentRunDataAggregator = new ParallelRunDataAggregator();
             this.concurrentManagerHandlerMap = new Dictionary<IProxyExecutionManager, ITestRunEventsHandler>();
 
-            for (int i = 0; i < this.concurrentManagerInstances.Length; i++)
+            foreach (var concurrentManager in this.concurrentManagerInstances)
             {
-                var concurrentManager = this.concurrentManagerInstances[i];
-
-                var parallelEventsHandler = new ParallelRunEventsHandler(
-                                                concurrentManager,
-                                                runEventsHandler,
+                var parallelEventsHandler = concurrentManager is ProxyExecutionManagerWithDataCollection ? 
```

Comment:
If it is possible that in given parallel session, datacollection is not enabled for all frameworks, then it is better to check if whether each concurentManager is PEM or PEMDC?